File: 30_traffic_lights_stm32_cpp/src/switch.cpp
#include "switch.h"

Switch::Switch(const Gpio gpio) : gpio(gpio) {}

void Switch::init()
{
    // here GPIOA is hardcoded
    __HAL_RCC_GPIOA_CLK_ENABLE();
    
    GPIO_InitTypeDef GPIO_InitStruct = {0};

    GPIO_InitStruct.Pin = gpio.pin;
    GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
    GPIO_InitStruct.Pull = GPIO_PULLUP;
    HAL_GPIO_Init(gpio.port, &GPIO_InitStruct);
}

bool Switch::is_on()
{
    return HAL_GPIO_ReadPin(gpio.port, gpio.pin) == GPIO_PIN_RESET;
}

File: 33_kf_esp32_embassy_rust/src/switch.rs
use esp_hal::gpio::{Input, InputConfig, InputPin, Pull};

pub struct Switch<'d> {
    input: Input<'d>,
}

impl<'d> Switch<'d> {
    pub fn new(pin: impl InputPin + 'd, pull: Pull) -> Self {
        let input_config = InputConfig::default().with_pull(pull);
        Self {
            input: Input::new(pin, input_config),
        }
    }
    pub fn is_on(&self) -> bool {
        self.input.is_low()
    }
}
